ambry-clustermap/src/main/java/com.github.ambry.clustermap/CompositeClusterManager.java

```diff
@@ -0,0 +1,272 @@
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+
+/**
+ * A cluster manager that is a wrapper over a {@link StaticClusterManager} instance and a {@link HelixClusterManager}
+ * instance and uses the {@link StaticClusterManager} as the source-of-truth. It relays events to both and checks for
+ * and reports inconsistencies in the views from the two underlying cluster managers.
+ */
+class CompositeClusterManager implements ClusterMap {
+  private final StaticClusterManager staticClusterManager;
+  private final HelixClusterManager helixClusterManager;
+  private final HelixClusterManagerMetrics helixClusterManagerMetrics;
+
+  /**
+   * Construct a CompositeClusterManager instance.
+   * @param staticClusterManager the {@link StaticClusterManager} instance to use as the source-of-truth.
+   * @param helixClusterManager the {@link HelixClusterManager} instance to use for comparison of views.
+   */
+  CompositeClusterManager(StaticClusterManager staticClusterManager, HelixClusterManager helixClusterManager) {
+    this.staticClusterManager = staticClusterManager;
+    this.helixClusterManager = helixClusterManager;
+    this.helixClusterManagerMetrics =
+        helixClusterManager != null ? helixClusterManager.helixClusterManagerMetrics : null;
+  }
+
+  @Override
+  public PartitionId getPartitionIdFromStream(InputStream stream) throws IOException {
+    DuplicatingInputStream duplicatingInputStream = new DuplicatingInputStream(stream);
+    duplicatingInputStream.mark(0);
+    PartitionId partitionIdStatic = staticClusterManager.getPartitionIdFromStream(duplicatingInputStream);
+    if (helixClusterManager != null) {
+      duplicatingInputStream.reset();
+      PartitionId partitionIdDynamic = helixClusterManager.getPartitionIdFromStream(duplicatingInputStream);
+      if (!partitionIdStatic.toString().equals(partitionIdDynamic.toString())) {
+        helixClusterManagerMetrics.getPartitionIdFromStreamMismatchCount.inc();
+      }
+    }
+    return partitionIdStatic;
+  }
+
+  /**
+   * Get writable partition ids from both the underlying {@link StaticClusterManager} and the underlying
+   * {@link HelixClusterManager}. Compare the two and if there is a mismatch, update a metric.
+   * @return a list of partition ids from the underlying {@link StaticClusterManager}.
+   */
+  @Override
+  public List<PartitionId> getWritablePartitionIds() {
+    List<PartitionId> staticWritablePartitionIds = staticClusterManager.getWritablePartitionIds();
+    if (helixClusterManager != null) {
+      Set<String> staticPartitionIdStrings = new HashSet<>();
+      for (PartitionId partitionId : staticWritablePartitionIds) {
+        staticPartitionIdStrings.add(partitionId.toString());
+      }
+
+      Set<String> helixPartitionIdStrings = new HashSet<>();
```

Comment:
This verification will never be removed, as the purpose of this class is to do this verification and update the metrics. We will at some point move to using the HelixClusterManager.